### PR TITLE
Drop old bootflag detection, rely on python3-parted

### PR DESCRIPTION
**Pull request by mika**
While at it rename "bootflag" into "boot flag" in human readable
strings.

Thanks: Darshaka Pathirana for bug report and debugging
Closes: grml/grml2usb#19

### CODE REVIEW

**Comment by jkirk** [comment]
The partition layout of the USB storage looks like this:

```
% sudo parted /dev/sdb p
Model: Ut165 USB2FlashStorage (scsi)
Disk /dev/sdb: 4041MB
Sector size (logical/physical): 512B/512B
Partition Table: msdos
Disk Flags: 

Number  Start   End     Size    Type     File system  Flags
 1      31.7kB  4007MB  4007MB  primary  fat16        boot
```

Run `grml2usb` with "`--dry-run --verbose`", "`--dry-run`" and without an argument (without `python3-parted` installed):

```
% sudo ./grml2usb --dry-run --verbose /srv/data/isos/grml64-full_testing_build3133.iso /dev/sdb1
2020-05-28 01:22:30,348 Executing grml2usb version v0.18.0-1-g5579b7b (git)
2020-05-28 01:22:30,348 Running in simulation mode as requested via option dry-run.
2020-05-28 01:22:30,421 Using ISO /srv/data/isos/grml64-full_testing_build3133.iso
2020-05-28 01:22:30,422 mount ['-o', 'loop,ro', '-t', 'iso9660'] /srv/data/isos/grml64-full_testing_build3133.iso /tmp/grml2usbrgjmihgy
2020-05-28 01:22:30,455 register_mountpoint(/tmp/grml2usbrgjmihgy)
2020-05-28 01:22:30,481 umount [] /tmp/grml2usbrgjmihgy
2020-05-28 01:22:30,518 unregister_mountpoint(/tmp/grml2usbrgjmihgy)
2020-05-28 01:22:30,518 Fatal: Could not import parted to verify boot flag on /dev/sdb1, please make sure python3-parted is installed.
2020-05-28 01:22:30,519 Exception:
Traceback (most recent call last):
  File "./grml2usb", line 509, in check_boot_flag
    import parted
ModuleNotFoundError: No module named 'parted'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./grml2usb", line 2162, in main
    install(iso, device)
  File "./grml2usb", line 1844, in install
    install_grml(iso_mountpoint, device)
  File "./grml2usb", line 1871, in install_grml
    check_boot_flag(device)
  File "./grml2usb", line 520, in check_boot_flag
    % device
VerifyException: Could not import parted to verify boot flag on /dev/sdb1, please make sure python3-parted is installed.

% sudo ./grml2usb --dry-run /srv/data/isos/grml64-full_testing_build3133.iso /dev/sdb1          
Executing grml2usb version v0.18.0-1-g5579b7b (git)
Running in simulation mode as requested via option dry-run.
Using ISO /srv/data/isos/grml64-full_testing_build3133.iso
Fatal: Could not import parted to verify boot flag on /dev/sdb1, please make sure python3-parted is installed.

% sudo ./grml2usb /srv/data/isos/grml64-full_testing_build3133.iso /dev/sdb1          
Executing grml2usb version v0.18.0-1-g5579b7b (git)
Using ISO /srv/data/isos/grml64-full_testing_build3133.iso
Fatal: Could not import parted to verify boot flag on /dev/sdb1, please make sure python3-parted is installed.
```
Looks good! :+1: 

Installed `python3-parted` and run again:

```
% sudo ./grml2usb /srv/data/isos/grml64-full_testing_build3133.iso /dev/sdb1        
Executing grml2usb version v0.18.0-1-g5579b7b (git)
Using ISO /srv/data/isos/grml64-full_testing_build3133.iso
Identified grml flavour "grml64-full_testing".
Copying files. This might take a while....
No /EFI/ubuntu/grub.cfg found inside EFI image, looks like Secure Boot support is missing.
Installing default MBR
Installing syslinux as bootloader
syslinux: /dev/sdb1: No such file or directory
Fatal: Error executing syslinux (either try --fat16 or use grub?)
Cleaning up before exiting...
```
A log message "Checking boot flag..." or so would be nice though.

Not sure yet, why syslinux fails. `/dev/sdb1` exists, syslinux is installed and the partition is `vfat`:

```
% dpkg -l syslinux
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name           Version                             Architecture Description
+++-==============-===================================-============-=======================================================
ii  syslinux       3:6.04~git20190206.bf6db5b4+dfsg1-1 amd64        collection of bootloaders (DOS FAT and NTFS bootloader)

% syslinux -v    
syslinux 6.04  Copyright 1994-2015 H. Peter Anvin et al

% sudo blkid -s TYPE -o value /dev/sdb1
vfat
```

The syslinux problem seems like a different problem, will do some more investigation to make sure.

**Comment by jkirk** [comment]
Removed the boot flag from the USB device:
```
(parted) p
Model: Ut165 USB2FlashStorage (scsi)
Disk /dev/sdb: 4041MB
Sector size (logical/physical): 512B/512B
Partition Table: msdos
Disk Flags:

Number  Start   End     Size    Type     File system  Flags
 1      31.7kB  4007MB  4007MB  primary  fat16
````

As I already copied the file I just wanted to set up the bootloader, so I added `--bootloader-only`.
Without `python3-parted` `grml2usb` failed as expected, but with it it seemed that no boot flag check was performed:
```
% sudo ./grml2usb --bootloader-only /srv/data/isos/grml64-full_testing_build3133.iso /dev/sdb1                                               
Executing grml2usb version v0.18.0-1-g5579b7b (git)
Using ISO /srv/data/isos/grml64-full_testing_build3133.iso
Identified grml flavour "grml64-full_testing".
No /EFI/ubuntu/grub.cfg found inside EFI image, looks like Secure Boot support is missing.
Installing default MBR
Installing syslinux as bootloader
syslinux: /dev/sdb1: No such file or directory
Fatal: Error executing syslinux (either try --fat16 or use grub?)
Cleaning up before exiting...
```
For the sake of it, I removed the `--bootloader-only` option, but the still the same behavior:
```
% sudo ./grml2usb /srv/data/isos/grml64-full_testing_build3133.iso /dev/sdb1                  
Executing grml2usb version v0.18.0-1-g5579b7b (git)
Using ISO /srv/data/isos/grml64-full_testing_build3133.iso
Identified grml flavour "grml64-full_testing".
Copying files. This might take a while....
No /EFI/ubuntu/grub.cfg found inside EFI image, looks like Secure Boot support is missing.
Installing default MBR
Installing syslinux as bootloader
syslinux: /dev/sdb1: No such file or directory
Fatal: Error executing syslinux (either try --fat16 or use grub?)
Cleaning up before exiting...
```

FTR, it takes quite some time (I would say minutes) between "No /EFI/ubuntu/grub.cfg found..." and "Installing default MBR".

File: grml2usb [hunk 1]
```diff
@@ -512,28 +512,13 @@ def check_boot_flag(device):
         if part is None:
             raise HodorException("parted could not find partition")
         if part.getFlag(parted.PARTITION_BOOT):
```
Comment (jkirk) [inline grml2usb:514]: It seems like this line does not do what we expect. But it does not matter if this line is true or false. As long as the `import parted` does not throw a `ImportError` this functions returns happily.

I think we need an `else` part with:
```
            logging.debug("bootflag is NOT enabled")
            raise VerifyException(
                "Device %s does not have the bootflag set. "
                "Please enable it to be able to boot." % device
            )
```

Comment (mika) [reply grml2usb:515]: Thanks for the analysis and the detailed report!

Took care of all your comments in https://github.com/grml/grml2usb/pull/29/commits/705a96b84994905fbc63e3926a5af785ac0c03bc - please review once again :)

**Comment by evgeni** [comment]
I remember writing this code, but I don't remember what it does, so this is a `pass` from me on the review, sorry.

**Comment by jkirk** [comment]
```
% sudo parted /dev/sdb p                                                                      
Model: Ut165 USB2FlashStorage (scsi)
Disk /dev/sdb: 4041MB
Sector size (logical/physical): 512B/512B
Partition Table: msdos
Disk Flags: 

Number  Start   End     Size    Type     File system  Flags
 1      31.7kB  4007MB  4007MB  primary  fat16
% sudo ./grml2usb --bootloader-only /srv/data/isos/grml64-full_testing_build3133.iso /dev/sdb1
Executing grml2usb version v0.18.0-1-g705a96b (git)
Using ISO /srv/data/isos/grml64-full_testing_build3133.iso
Checking for boot flag
Fatal: Device /dev/sdb1 does not have the bootflag set. Please enable it to be able to boot.

% sudo ./grml2usb /srv/data/isos/grml64-full_testing_build3133.iso /dev/sdb1                  
Executing grml2usb version v0.18.0-1-g705a96b (git)
Using ISO /srv/data/isos/grml64-full_testing_build3133.iso
Checking for boot flag
Fatal: Device /dev/sdb1 does not have the bootflag set. Please enable it to be able to boot.


% sudo parted /dev/sdb  
GNU Parted 3.2
Using /dev/sdb
Welcome to GNU Parted! Type 'help' to view a list of commands.
(parted) toggle 1 boot
(parted) p                                                                
Model: Ut165 USB2FlashStorage (scsi)
Disk /dev/sdb: 4041MB
Sector size (logical/physical): 512B/512B
Partition Table: msdos
Disk Flags: 

Number  Start   End     Size    Type     File system  Flags
 1      31.7kB  4007MB  4007MB  primary  fat16        boot

(parted) q                                                                
Information: You may need to update /etc/fstab.

% sudo ./grml2usb /srv/data/isos/grml64-full_testing_build3133.iso /dev/sdb1
Executing grml2usb version v0.18.0-1-g705a96b (git)
Using ISO /srv/data/isos/grml64-full_testing_build3133.iso
Checking for boot flag
Identified grml flavour "grml64-full_testing".
Copying files. This might take a while....
No /EFI/ubuntu/grub.cfg found inside EFI image, looks like Secure Boot support is missing.
[...Stuck for minute(s)... ]
Synching data (this might take a while)
[...Stuck for minute(s) again... ]
Installing default MBR
Installing syslinux as bootloader
syslinux: /dev/sdb1: No such file or directory
Fatal: Error executing syslinux (either try --fat16 or use grub?)
Cleaning up before exiting...

% sudo ./grml2usb --bootloader-only /srv/data/isos/grml64-full_testing_build3133.iso /dev/sdb1
Executing grml2usb version v0.18.0-1-g705a96b (git)
Using ISO /srv/data/isos/grml64-full_testing_build3133.iso
Checking for boot flag
Identified grml flavour "grml64-full_testing".
No /EFI/ubuntu/grub.cfg found inside EFI image, looks like Secure Boot support is missing.
Synching data (this might take a while)
[...Stuck for minute(s)... ]
Installing default MBR
Installing syslinux as bootloader
syslinux: /dev/sdb1: No such file or directory
Fatal: Error executing syslinux (either try --fat16 or use grub?)
Cleaning up before exiting...
```

So far it looks good. I have to test whether this USB storage actually boots (aka syslinux error). But for now I tend to ACK this PR. :+1:

**Comment by jkirk** [comment]
OK. Sadly this USB device does not boot on my other notebook. New issue incoming... but I have to investigate first.

**Comment by mika** [comment]
> OK. Sadly this USB device does not boot on my other notebook. New issue incoming... but I have to investigate first.

I'd assume that's related to:

```
Installing syslinux as bootloader
syslinux: /dev/sdb1: No such file or directory
Fatal: Error executing syslinux (either try --fat16 or use grub?)
```

?

**Pull request merged**